components/sidebar/sidebar_menu.tsx

```diff
@@ -0,0 +1,143 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+import React from 'react';
+import {Tooltip} from 'react-bootstrap';
+import classNames from 'classnames';
+
+import OverlayTrigger from 'components/overlay_trigger';
+import MenuWrapper from 'components/widgets/menu/menu_wrapper';
+import Menu from 'components/widgets/menu/menu';
+
+const MENU_BOTTOM_MARGIN = 80;
+
+type Props = {
+    id: string;
+    tooltipText: string;
+    buttonAriaLabel: string;
+    ariaLabel: string;
+    refCallback?: (ref: SidebarMenu) => void;
+};
+
+type State = {
+    isMenuOpen: boolean;
+    openUp: boolean;
+    width: number;
+};
+
+export default class SidebarMenu extends React.PureComponent<Props, State> {
+    menuRef?: Menu;
+    menuButtonRef: React.RefObject<HTMLButtonElement>;
+    isLeaving: boolean;
+
+    constructor(props: Props) {
+        super(props);
+
+        this.state = {
+            isMenuOpen: false,
+            openUp: false,
+            width: 0,
+        };
+
+        this.menuButtonRef = React.createRef();
+        this.isLeaving = false;
+    }
+
+    // TODO: Temporary code to keep the menu in place while scrolling
+    componentDidMount() {
+        const scrollbars = document.querySelectorAll('#SidebarContainer .SidebarNavContainer .scrollbar--view');
+        if (scrollbars && scrollbars[0]) {
+            scrollbars[0].addEventListener('scroll', this.setMenuPosition);
+        }
+    }
+
+    componentWillUnmount() {
+        const scrollbars = document.querySelectorAll('#SidebarContainer .SidebarNavContainer .scrollbar--view');
+        if (scrollbars && scrollbars[0]) {
+            scrollbars[0].removeEventListener('scroll', this.setMenuPosition);
+        }
+    }
+
+    refCallback = (ref: Menu) => {
```

Comment:
Similar to my other comment, could more of this be done in `handleMenuToggle` so we position the menu right before its shown instead of hijacking the ref for it?

Reply:
See above comment.